tests/System.Text.Primitives.Tests/EncodingPerfTestsCoreCLR.cs

```diff
@@ -0,0 +1,180 @@
+                    output = Encoding.Convert(utf8, utf8, utf8Buffer);
+            }
+        }
+
+        [Benchmark]
+        [InlineData(DataLength, 0x0, Utf8OneByteLastCodePoint)]
+        [InlineData(DataLength, Utf8OneByteLastCodePoint + 1, Utf8TwoBytesLastCodePoint)]
+        [InlineData(DataLength, Utf8TwoBytesLastCodePoint + 1, Utf8ThreeBytesLastCodePoint)]
+        [InlineData(DataLength, Utf16HighSurrogateFirstCodePoint, Utf16LowSurrogateLastCodePoint)]
+        [InlineData(DataLength, 0x0, Utf8ThreeBytesLastCodePoint)]
+        [InlineData(DataLength, 0, 0, SpecialTestCases.AlternatingASCIIAndNonASCII)]
+        [InlineData(DataLength, 0, 0, SpecialTestCases.MostlyASCIIAndSomeNonASCII)]
+        public void EncodeFromUtf16toUtf8CoreCLR(int length, int minCodePoint, int maxCodePoint, SpecialTestCases special = SpecialTestCases.None)
+        {
+            string inputString = GenerateStringData(length, minCodePoint, maxCodePoint, special);
+            char[] characters = inputString.AsSpan().ToArray();
+            Encoding utf8 = Encoding.UTF8;
+            int utf8Length = utf8.GetByteCount(characters);
+            var utf8Buffer = new byte[utf8Length];
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                    utf8.GetBytes(characters, 0, characters.Length, utf8Buffer, 0);
+            }
+        }
+
+        [Benchmark]
+        [InlineData(DataLength, 0x0, Utf8OneByteLastCodePoint)]
+        [InlineData(DataLength, Utf8OneByteLastCodePoint + 1, Utf8TwoBytesLastCodePoint)]
+        [InlineData(DataLength, Utf8TwoBytesLastCodePoint + 1, Utf8ThreeBytesLastCodePoint)]
+        [InlineData(DataLength, Utf16HighSurrogateFirstCodePoint, Utf16LowSurrogateLastCodePoint)]
+        [InlineData(DataLength, 0x0, Utf8ThreeBytesLastCodePoint)]
+        [InlineData(DataLength, 0, 0, SpecialTestCases.AlternatingASCIIAndNonASCII)]
+        [InlineData(DataLength, 0, 0, SpecialTestCases.MostlyASCIIAndSomeNonASCII)]
+        public void EncodeFromUtf32toUtf8CoreCLR(int length, int minCodePoint, int maxCodePoint, SpecialTestCases special = SpecialTestCases.None)
+        {
+            string inputString = GenerateStringData(length, minCodePoint, maxCodePoint, special);
+            char[] characters = inputString.AsSpan().ToArray();
+            Encoding utf32 = Encoding.UTF32;
+            Encoding utf8 = Encoding.UTF8;
+            int utf32Length = utf32.GetByteCount(characters);
+            var utf32Buffer = new byte[utf32Length];
+            utf32.GetBytes(characters, 0, characters.Length, utf32Buffer, 0);
+
+            int utf8Length = utf8.GetByteCount(characters);
+            var output = new byte[utf8Length];
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                    output = Encoding.Convert(utf32, utf8, utf32Buffer);
```

Comment:
This implementation is not really consistent with what we do in the primitives. In order for the performance tests to be meaningful, I would pick the fastest way via the CoreCLR implementation to compare to our direct implementation. The most efficient way in CoreCLR would be to pre-allocate the buffers (not part of the performance test since we do the same with Span), then call the necessary decoders / encoders to get where you want. If there is no direct path, use as few links in the chain as you can. Most cases,  you can go through UTF-16.